app/src/main/java/com/github/libretube/ui/tools/SleepTimer.kt

```diff
@@ -6,52 +6,67 @@ import android.os.Looper
 import android.os.Process
 import androidx.core.os.postDelayed
 import com.github.libretube.R
-import com.github.libretube.constants.PreferenceKeys
-import com.github.libretube.helpers.PreferenceHelper
 import com.github.libretube.ui.activities.MainActivity
 import com.google.android.material.snackbar.Snackbar
+import java.util.Timer
+import kotlin.concurrent.scheduleAtFixedRate
 
 object SleepTimer {
     private val handler = Handler(Looper.getMainLooper())
     private const val REACTION_INTERVAL = 5L
+    private const val TIMER_DELAY = 1000L
+    private lateinit var timer: Timer
+    var timeLeftMillis: Long = 0L
 
     /**
      * Kill the app after showing a warning after a certain amount of time
-     * @param context This must not be the applicationContext!
+     *
+     * @param context This must not be the applicationContext, but an activity context!
      */
-    fun setup(context: Context) {
-        if (!PreferenceHelper.getBoolean(PreferenceKeys.SLEEP_TIMER, false)) return
+    fun setup(context: Context, delayInMinutes: Long) {
+        if (delayInMinutes == 0L) return
 
-        val breakReminderPref = PreferenceHelper.getString(
-            PreferenceKeys.SLEEP_TIMER_DELAY,
-            ""
-        ).ifEmpty { return }
+        timeLeftMillis = delayInMinutes * 60 * 1000
 
-        handler.postDelayed(breakReminderPref.toLong() * 60 * 1000) {
-            var killApp = true
-            val mainActivity = context as? MainActivity ?: return@postDelayed
-            val snackBar = Snackbar.make(
-                mainActivity.binding.root,
-                R.string.take_a_break,
-                Snackbar.LENGTH_INDEFINITE
-            )
-                .setAction(R.string.cancel) {
-                    killApp = false
-                }
-            snackBar.show()
-            for (i in 0..REACTION_INTERVAL) {
-                handler.postDelayed(i * 1000) {
-                    val remainingTime = " (${REACTION_INTERVAL - i})"
-                    snackBar.setText(context.getString(R.string.take_a_break) + remainingTime)
-                }
+        timer = Timer()
+        timer.scheduleAtFixedRate(TIMER_DELAY, TIMER_DELAY) {
+            timeLeftMillis -= TIMER_DELAY
+            if (timeLeftMillis == 0L) showTimerEndedSnackBar(context)
+        }
+    }
+
+    /**
+     * Disable the scheduled sleep timer
+     */
+    fun disableSleepTimer() {
+        if (this::timer.isInitialized) timer.cancel()
```

Comment:
The timer could be a nullable variable instead, that way this couls be simplified to timer?.cancel().